pkg/ipamd/datastore/data_store.go

```diff
@@ -1465,22 +1471,27 @@ func (ds *DataStore) CheckFreeableENIexists() bool {
 }
 
 // NormalizeCheckpointDataByPodVethExistence will normalize checkpoint data by removing allocations that do not have a corresponding pod veth.
-// This should not happen unless container runtimes have bugs that failed to invoke the cmdDel for died pods.
-// We use this reconciliation as a safety mechanism during transition from CRI to state file.
+// This can happen if pods are deleted while IPAMD is inactive.
 func (ds *DataStore) normalizeCheckpointDataByPodVethExistence(checkpoint CheckpointData) (CheckpointData, error) {
 	hostNSLinks, err := ds.netLink.LinkList()
 	if err != nil {
 		return CheckpointData{}, err
 	}
 	var validatedAllocations []CheckpointEntry
+	var staleAllocations []CheckpointEntry
 	for _, allocation := range checkpoint.Allocations {
 		if err := ds.validateAllocationByPodVethExistence(allocation, hostNSLinks); err != nil {
-			ds.log.Warnf("ignore IP allocation for %v:%v,%v due to %v", allocation.ContainerID, allocation.IPv4, allocation.IPv6, err)
+			ds.log.Warnf("stale IP allocation for ID(%v): IPv4(%v), IPv6(%v) due to %v", allocation.ContainerID, allocation.IPv4, allocation.IPv6, err)
+			staleAllocations = append(staleAllocations, allocation)
 		} else {
 			validatedAllocations = append(validatedAllocations, allocation)
 		}
 	}
 	checkpoint.Allocations = validatedAllocations
+	// Stale allocations may have dangling IP rules that need cleanup
+	if len(staleAllocations) > 0 {
+		ds.PruneStaleAllocations(staleAllocations)
```

Comment:
We don't move the IPAMD/CNI delete logic, we need to cleanup the backing store too right?

Reply:
The backing store is cleaned up just by writing from the in-memory datastore after init